spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/DockerCommandBuilder.java

```diff
@@ -61,24 +64,47 @@ public DockerCommandBuilder(String dockerNetwork, boolean deleteContainerOnExit)
 	}
 
 	@Override
-	public String[] buildExecutionCommand(AppDeploymentRequest request, Map<String, String> appInstanceEnv,
-										  Optional<Integer> appInstanceNumber) {
-		List<String> commands = addDockerOptions(request, appInstanceEnv, appInstanceNumber);
+	public int getPortSuggestion(LocalDeployerProperties localDeployerProperties) {
+		return ThreadLocalRandom.current().nextInt(localDeployerProperties.getDocker().getPortRange().getLow(),
+				localDeployerProperties.getDocker().getPortRange().getHigh());
+	}
+
+	@Override
+	public URL getBaseUrl(String deploymentId, int index, int port) {
+		try {
+			return new URL("http", String.format("%s-%d", deploymentId, index), port, "");
+		}
+		catch (Exception e) {
+			throw new IllegalArgumentException(e);
+		}
+	}
+
+	@Override
+	public ProcessBuilder buildExecutionCommand(AppDeploymentRequest request, Map<String, String> appInstanceEnv, String deployerId,
+			Optional<Integer> appInstanceNumber, LocalDeployerProperties localDeployerProperties,
+			Optional<DebugAddress> debugAddressOption) {
+
+		appInstanceEnv.put("deployerId", deployerId);
+		List<String> commands = addDockerOptions(request, appInstanceEnv, appInstanceNumber, localDeployerProperties, debugAddressOption);
 		commands.addAll(request.getCommandlineArguments());
 		logger.debug("Docker Command = " + commands);
-		return commands.toArray(new String[0]);
+		return new ProcessBuilder(Arrays.asList(AbstractLocalDeployerSupport.windowsSupport(commands.toArray(new String[0]))));
 	}
 
 	private List<String> addDockerOptions(AppDeploymentRequest request, Map<String, String> appInstanceEnv,
-										  Optional<Integer> appInstanceNumber) {
+			Optional<Integer> appInstanceNumber, LocalDeployerProperties localDeployerProperties,
+			Optional<DebugAddress> debugAddressOption) {
+
 		List<String> commands = new ArrayList<>();
 		commands.add("docker");
 		commands.add("run");
+
 		if (StringUtils.hasText(this.dockerNetwork)) {
 			commands.add("--network");
 			commands.add(this.dockerNetwork);
 		}
-		if (this.deleteContainerOnExit) {
+
+		if (this.deleteContainerOnExit && localDeployerProperties.getDocker().isDeleteContainerOnExit()) {
```

Comment:
I'm not following this. Should it be `||` ?

Reply:
The logic is that you can disable container-deletion-on-exit for all containers (e.g. `this.deleteContainerOnExit`) or for a singe app container (`localDeployerProperties.getDocker().isDeleteContainerOnExit()`). 

In current implementation the first is configured at [local deployer boot time](https://github.com/spring-cloud/spring-cloud-deployer-local/blob/master/spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/AbstractLocalDeployerSupport.java#L106-L107) e.g. you app you this with DataFlow's `SPRING_CLOUD_DATAFLOW_TASK_PLATFORM_LOCAL_ACCOUNTS_DEFAULT_DOCKER_DELETE_CONTAINER_ON_EXIT`  or Skipper's `SPRING_CLOUD_SKIPPER_SERVER_PLATFORM_LOCAL_ACCOUNTS_DEFAULT_DOCKER_DELETE_CONTAINER_ON_EXIT` properties. 
Then you can use the pipeline deployment property to set the desired behavior for a particular app at the time of deployment. 

In other words you can set the deletion to true by default, but have the choice to alter this when you deploy the pipeline, which is handy to check the logs of some failing app without having to  restart/reconfigure the SCDF&Skipper.

Reply:
So if I ask for a specific container to be deleted, I expect that would override the default setting

Reply:
indeed

Reply:
this.deleteOnExit=false
this.deployerProperties.getDocker().isDeleteContainerOnExit()=true
then `true && false == false`. So no `--rm`.  
What am I missing.

Reply:
Note that `this.deployerProperties.getDocker().isDeleteContainerOnExit()=true` is per application **not** per stream/pipeline! So in this case the `no --rm` means that all other apps in the pipeline will be deleted but this one will be kept so you check the logs or inspect its environment. 
I've already used it for troubleshooting a failing app. IMO It is easier if you try it. Use `watch docker ps -a` to monitor the behavior of the apps in the pipeline after you deploy or un-deploy it.

Reply:
You can think of `this.deleteOnExit` as policy for the entire pipeline, while the `this.deployerProperties.getDocker().isDeleteContainerOnExit()` is per single app in that pipeline.
By default the this.deleteOnExit is true because in general case you don't want to keep all containers for all pipelines. Thinking about this perhaps `this.deleteOnExit=false` e.g. keep all containers for all pipelines is never desired? Perhaps using the this.deployerProperties.getDocker().isDeleteContainerOnExit() for cherry pinking what containers to stay would be enough?

Reply:
I will remove the `this.deleteOnExit`

Reply:
https://github.com/spring-cloud/spring-cloud-deployer-local/pull/198/files#diff-9984d89909b44a5e9c80671c6cafb573R105-R107